### PR TITLE
Implement proxy templates

### LINKED ISSUE #6785: Create tsh subcommand to be used within ProxyCommand.
In [RFD 21: Cluster Routing](https://github.com/gravitational/teleport/blob/master/rfd/0021-cluster-routing.md) `tsh ssh -J` support was added to support `tsh` directly connecting to a node within a leaf cluster (to bypass latency issues). To support this, `tsh` was taught how to dynamically request and obtain cluster specific certificates similar to `tsh login leafCluster`.

This work for `tsh`, but does not for `ssh`. To support this functionality, a `tsh` subcommand will have to be created that can perform these operations behind the scene and called from within a `ProxyCommand`.

An example of what this subcommand would look like is below.

```
Host *.example.com
    HostName %h
    Port 3022
    ProxyCommand tsh proxy-jump %r %h clusterName
```

### LINKED ISSUE #10488: Simplify OpenSSH client integration with Teleport
If you have many leaf clusters with Teleport today, your `~/.ssh/config` grows in complexity each time you add a cluster. Furthermore, if you use `-J` to directly connect to a leaf cluster, you have to loop over all clusters and log in to each one everytime you add a leaf cluster.

Simplify this process for users so they can add a simple `~/.ssh/config` that can be used across multiple leaf clusters and reduces maintenance complexity.

RFD: https://github.com/gravitational/teleport/pull/11528.

### PR DESCRIPTION
**Pull request by r0mant**
Implements [RFD 62: Proxy templates](https://github.com/gravitational/teleport/blob/master/rfd/0062-tsh-proxy-template.md) (with some minor amendments to the RFD).

Closes https://github.com/gravitational/teleport/issues/10488, closes https://github.com/gravitational/teleport/issues/6785.

### CODE REVIEW

**Comment by ptgott** [comment]
Since this PR includes code changes as well as docs additions, @r0mant please let me know when the code is stable enough for me to review the docs changes.

Also, [this pending PR](https://github.com/gravitational/teleport/pull/12183) changes the OpenSSH guide into a step-by-step guide. We might want to wait until that PR is merged, then add a separate guide for using proxy templates in an OpenSSH environment. In any case, I'll give this a more detailed review once it's ready.

File: api/client/webclient/webclient.go
Comment (nklaassen) [inline api/client/webclient/webclient.go:459]: this logic is pretty fun, are we effectively duplicating it from somewhere else? Hard to imagine this is the only place we need to determine the SSH proxy address, would probably be kind of hard to extract but I'd hate to see it get out of sync if it is

File: tool/tsh/tshconfig_test.go
Comment (nklaassen) [inline tool/tsh/tshconfig_test.go:222]: Can we test some tricker replacements? Something with numbers or letters next to the replacement maybe so you need braces, `${1}1-x${2}.example.com:3080` or something

File: tool/tsh/proxy.go
Comment (smallinsky) [inline tool/tsh/proxy.go:104]: The JumpHost address provided by a user can point to "normal" (not TLSRouting multiplexed)  `SSHProxyListenPort == 3023` teleport  proxy ssh service. In that case this ping call (`webclient.Find(ProxyAddr: tc.JumpHosts[0].Addr.Addr`) will always fail.

Comment (r0mant) [reply tool/tsh/proxy.go:104]: A user is supposed to point it to web proxy address. I updated the RFD to mention this as well in this PR. The SSH proxy address will be determined from the ping endpoint.

File: api/client/webclient/webclient.go
Comment (Joerger) [inline api/client/webclient/webclient.go:442]: nit: Should the default port be `ps.SSH.WebListener.Port` if non-empty?

As a user, I'd expect the following config to produce `proxy.example.com:443`, but this would currently produce `proxy.example.com:3080`.

```
proxy_service:
    public_addr: ["proxy.example.com"]
    web_listen_addr: 0.0.0.0:443
```

Note: this is a potential problem across the board that I've noticed, so it's somewhat unrelated to this PR.

File: docs/pages/server-access/guides/openssh.mdx
Comment (Joerger) [inline docs/pages/server-access/guides/openssh.mdx:404]: nit: It might be worth mentioning that the "potential latency issues" are only relevant in regionally distributed clusters, if I'm understanding correctly.
```suggestion
However, in regionally distributed clusters where leaf proxies are reachable by
SSH clients, you might prefer to connect directly through the leaf proxies to avoid
potential latency issues of the connection being routed through the root cluster.
```

File: tool/tsh/tshconfig.go
Comment (Joerger) [reply tool/tsh/tshconfig.go:82]: Yeah it should be the other way around. I think order doesn't affect `ExtraHeaders` but I'll flip both regardless.

**Review by smallinsky — APPROVED** [review]

**Comment by Joerger** [comment]
@ptgott This PR is ready for Docs review when you get a chance

**Review by nklaassen — APPROVED** [review]

**Pull request merged**
